core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,199 @@
+ */
+
+package org.apache.spark
+
+import java.util.{Timer, TimerTask}
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.{RpcCallContext, RpcEnv, ThreadSafeRpcEndpoint}
+import org.apache.spark.scheduler.{LiveListenerBus, SparkListener, SparkListenerStageCompleted}
+
+/**
+ * A coordinator that handles all global sync requests from BarrierTaskContext. Each global sync
+ * request is generated by `BarrierTaskContext.barrier()`, and identified by
+ * stageId + stageAttemptId + barrierEpoch. Reply all the blocking global sync requests upon
+ * received all the requests for a group of `barrier()` calls. If the coordinator doesn't collect
+ * enough global sync requests within a configured time, fail all the requests due to timeout.
+ */
+private[spark] class BarrierCoordinator(
+    timeout: Int,
+    listenerBus: LiveListenerBus,
+    override val rpcEnv: RpcEnv) extends ThreadSafeRpcEndpoint with Logging {
+
+  private val timer = new Timer("BarrierCoordinator barrier epoch increment timer")
+
+  private val listener = new SparkListener {
+    override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
+      val stageInfo = stageCompleted.stageInfo
+      // Remove internal data from a finished stage attempt.
+      cleanupSyncRequests(stageInfo.stageId, stageInfo.attemptNumber)
+      barrierEpochByStageIdAndAttempt.remove((stageInfo.stageId, stageInfo.attemptNumber))
+    }
+  }
+
+  // Epoch counter for each barrier (stage, attempt).
+  private val barrierEpochByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), AtomicInteger]
+
+  // Remember all the blocking global sync requests for each barrier (stage, attempt).
+  private val syncRequestsByStageIdAndAttempt =
+    new ConcurrentHashMap[(Int, Int), ArrayBuffer[RpcCallContext]]
+
+  override def onStart(): Unit = {
+    super.onStart()
+    listenerBus.addToStatusQueue(listener)
+  }
+
+  /**
+   * Get the array of [[RpcCallContext]]s that correspond to a barrier sync request from a stage
+   * attempt.
+   */
+  private def getOrInitSyncRequests(
+      stageId: Int,
+      stageAttemptId: Int,
+      numTasks: Int = 0): ArrayBuffer[RpcCallContext] = {
+    val requests = syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      new ArrayBuffer[RpcCallContext](numTasks))
+    if (requests == null) {
+      syncRequestsByStageIdAndAttempt.get((stageId, stageAttemptId))
+    } else {
+      requests
+    }
+  }
+
+  /**
+   * Clean up the array of [[RpcCallContext]]s that correspond to a barrier sync request from a
+   * stage attempt.
+   */
+  private def cleanupSyncRequests(stageId: Int, stageAttemptId: Int): Unit = {
+    val requests = syncRequestsByStageIdAndAttempt.remove((stageId, stageAttemptId))
+    if (requests != null) {
+      requests.clear()
```

Comment:
is this needed? when we call `syncRequestsByStageIdAndAttempt.remove((stageId, stageAttemptId))`, the array buffer becomes dangling and will be GCed.

Reply:
This is just to be safe, in case the requests are held in other places, we can still GC the `RpcCallContext`s

Reply:
Agree with @cloud-fan that this is not necessary. It only explicitly clears the ArrayBuffer object instead of the contexts.